### DISCUSSION TITLE
separation of concerns in routers

### DISCUSSION THREAD

**Discussion by RamonVigdor** [fastapi-users/fastapi-users#926] [Q&A] [answered]
Hi all, at least at first glance, it seems to me that routers mix auth logic with router logic. See e.g. below from oauth.py get_oauth_router:
If we like to return a RedirectResponse instead of the OAuth2AuthorizeResponse (or do any other route related change), we need to rewrite this router (or am I missing something here?). But this router includes lower-level auth logic. Rewriting this would mean manually tracking updates to any future auth-related logic changes in fastapi-users. Would it not be a good idea to separate the concerns and allow the user to build arbitrary routes that internally call auth methods from fastapi-users? The above holds, of course, only in case I am not missing something totally trivial making my point pointless ;-) Please enlighten me     

@router.get(
        "/authorize",
        name=f"oauth:{oauth_client.name}.{backend.name}.authorize",
        response_model=OAuth2AuthorizeResponse,
    )
    async def authorize(
        request: Request, scopes: List[str] = Query(None)
    ) -> OAuth2AuthorizeResponse:
        if redirect_url is not None:
            authorize_redirect_url = redirect_url
        else:
            authorize_redirect_url = request.url_for(callback_route_name)

        state_data: Dict[str, str] = {}
        state = generate_state_token(state_data, state_secret)
        authorization_url = await oauth_client.get_authorization_url(
            authorize_redirect_url,
            state,
            scopes,
        )

        return OAuth2AuthorizeResponse(authorization_url=authorization_url)

**Comment by frankie567** [comment, marked as answer]
Most of the actual logic actually lives in the [UserManager](https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/manager.py). For most routes, the core logic should be reusable from this class.

However, you're right on the example you show, the `/authorize` may have too many core logic inside that should live in the manager.
